torch/_dynamo/guards.py

```diff
@@ -514,6 +524,14 @@ def DEFAULT_DEVICE(self, guard: Guard):
             guard, [f"utils_device.CURRENT_DEVICE == {m.CURRENT_DEVICE!r}"]
         )
 
+    def BACKEND_MATCH(self, guard: Guard):
```

Comment:
We don't want more than 1 of these. We also need an invariant to ensure we don't somehow cross wires and try to accumulate guards across different backends for the same guard fn.

I would change this to be a flag that says "Yea, we need a backend match w/ backend FOO" and when we go to produce and install the check_fn, that's where we add this to the very front of the guard set. If another flag is raised, where the backend does not match FOO in our example, we need to raise.

Reply:
We also need to be very careful that identity is good enough here - there's a reason we were toying with the idea of "`'eager'` is special" - because it avoids a whole class of cache destroying problems where a seemingly mismatched identity, for a functionally identical backend, means we never cache hit.

Reply:
> We also need an invariant to ensure we don't somehow cross wires and try to accumulate guards across different backends for the same guard fn.

It seems to me the invariant is ready: BACKEND_GUARD is added to the tracing_context in OutputGraph's constructor. So it is 1-1 mapping with the root instruction translator. So for a frame, the first time dynamo encounter this frame, it'll write this guard into the check_fn when dynamo successfully compiles a graph. Later on, if we encountered it again but with a different backend, we'll do a recompilation.

> We also need to be very careful that identity is good enough

Yes, I agree. Jason's comment mentions two cases. And we don't have a good way for lambda problem right now and the `== ` can address the wrapper problem for the wrappers that we have now.

Reply:
This will add BACKEND_GUARD as many times as it encounters it. As I said above, there should only be a single one. And that single one should be at the very fore of the check_fn. Please replace this code to not append guard code, but to set a flag instead, which then instructs us how to write the code to the front of check_fn.

Reply:
Ok, let me try to digest a bit. So you're suggesting that we should keep an 1-1 mapping between backend and original code object and raise exceptions if the code object is compiled with a different backend. Specifically, we keep the backend that dynamo uses to compile this code object for the first time. If later on, any torch.compile is applied on this code object again but with a different backend, we should raise an exception. Is this the proposal?

I have some concerns about this single-cache design: 
1. @jansel 's the lambda and ephemeral wrapper case would not be possible: in these cases, we expect different lambdas or wrappers to share the optimized code object instead of raising on exception. Or, if we somehow can make sharing possible (e.g. for wrapper, we checkwith `==` instead of `is`), I would argue that we could do the same for current design to avoid repeated compilation.
2. We still need to figure out how to implement a separate "eager" cache for torch.cond to support automatic dynamo inspection. 
3. How do we scale this design to compile configurations beyond backend? e.g. fullgraph, capture_dynamic_output_shape_ops etc. Even for the same "eager" backend, these configuration changes may need a separate cache to avoid incorrect behaviors.

Reply:
> Ok, let me try to digest a bit. So you're suggesting that we should keep an 1-1 mapping between backend and original code object and raise exceptions if the code object is compiled with a different backend. Specifically, we keep the backend that dynamo uses to compile this code object for the first time. If later on, any torch.compile is applied on this code object again but with a different backend, we should raise an exception. Is this the proposal?

No, this is not the proposal, at all. Let me start again.

With the code as you have written it, your `BACKEND_MATCH` will add code over and over to the guards, for every BACKEND_MATCH we accumulate during trace. This is wasteful. Imagine stumbling upon a `cond` during trace incurs a `BACKEND_MATCH` - and let's say we have 4 conds in the traced program. With how this PR is written, we will have 4 identical BACKEND_MATCH. This is wasteful, and also fails to protect a useful invariant (that, which, for a given traced frame, the backend in BACKEND_MATCH is always the same).

**Current code:**

A BACKEND_MATCH that adds backend matching code for every BACKEND_MATCH in the frame we are tracing

**My proposal:**

A BACKEND_MATCH that sets a flag upon discovery - subsequent accumulations of BACKEND_MATCH just check that that the backend is the same as found in earlier BACKEND_MATCH's accumulated during this frame. If this flag is set, we add a single backend comparison to the front of the check fn

**Proposed test:**

Write code that accumulates K BACKEND_MATCHs, where K > 1, and then other guards as well

First compile: BACKEND_MATCH + other guards
Second compile: Recompile due to backend change AND tensor change

Assert that (1) there is only 1 backend comparison in the guard fn (despite having K accumulations) and (2) that the failure was because of backend mismatch (proving this guard was moved to the very front of the check fn)

Reply:
> Imagine stumbling upon a cond during trace incurs a BACKEND_MATCH - and let's say we have 4 conds in the traced program. With how this PR is written, we will have 4 identical BACKEND_MATCH.

No, we'll only create 1 BACKEND_MATCH guard and match it 3 times. See below example based on current PR and the dynamo logs:
```python
import torch
from torch._dynamo.testing import CompileCounter

cnt = CompileCounter()
x = torch.randn([3, 4])
def cond(x):
    def k(x):
        return x.sin(), x.cos()
    return torch.compile(k, backend=cnt)(x)

def g(x):
    return cond(x), cond(x), cond(x), cond(x)

g(x)
print(f"frame_count:{cnt.frame_count} op_count:{cnt.op_count}")
```
logs: https://www.internalfb.com/phabricator/paste/view/P811076827

Reply:
> This can happen when we have graph breaks

This is a new frame, so will get a new check_fn.

> So is this the case that you refer to? 

No

> No, we'll only create 1 BACKEND_MATCH guard and match it 3 times

This is incidentally the case because you invoke it once, but the function is written in a way that does not protect this invariant. We can write it better. I explained how above.

Reply:
In the case of cond, that's the exact pattern, right? I don't see it as an incident. Can you show me an example of multiple BACKEND_MATCH guards in a program? I haven't been able to come up with one...

Reply:
My point is, anyone can go and add it. This PR doesnt make that mistake, but as the function is written, it can happen. Just do `add_guard(GuardBuilder.BACKEND_MATCH)`, and bam, there's your bug. A well written version of this guard makes that impossible, and is not really any more complicated than what you have today, AND has the added bonus of moving this guard to the forefront of the check_fn.

Reply:
Yeah, I can buy the argument that someone might do a random `add_guard(GuardBuilder.BACKEND_MATCH)` somewhere. I'll wait for more comment and modify it accordingly. Thanks!

Reply:
Voz, I don't understand the point you are making here. Guards are tossed in a set so they will get deduped anyway. And we don't actually have any precedent for what you suggested; all of the "do it once" guards in OutputGraph are done in the same way this guard was done

Reply:
No - those are global state guards - for things we cannot discover lazily / AKA need ambient global state protection. DETERMINISTIC_ALGORITHMS, GRAD_MODE, DEFAULT_DEVICE, TORCH_FUNCTION_STATE

This is a little different. 

1) There is no reason to accrue this guard if we did not hit upon a higher order op that always invoke PT2, even not under a compile region. My understanding was that this was entirely for `cond` operators.

If the goal is to have a per-backend cache, and we want this available generally, then this isn't the way to do it IMO.

2) This guard should be at the front of the other guards, I dont think we have a good mechanism for this other than the one I proposed.

IDC too much though, I can open a PR with my proposed changes if we decide to land this. It's not blocking, just less than ideal.

Reply:
For 1, I want add that this pr is more than `entirely for cond`. In fact, I've triggered an silent failure while developing this PR. 

For example, we have [tests] that rely on first compile a function with one backend then compile a test with another backend ("aot_eager" and "inductor").https://github.com/pytorch/pytorch/blob/e0f1fe102ac85bcf3b7b01e50fa56df1bcbbfef1/torch/testing/_internal/optests/compile_check.py#L47-L48

Before this PR, we never thought about the second torch.compile with "inductor" backend doesn't trigger compilation due to lacking of a BACKEND_MATCH guard.